File: CourseSchedule.java
import java.util.ArrayList;
import java.util.HashSet;

public class CourseSchedule {

    public static void main(String[] args) {
        int[][] prerequisites = new int[][] {{0,1}};
        System.out.println(canFinish(2, prerequisites));
    }

    public static boolean canFinish(int numCourses, int[][] prerequisites) {
        HashSet<Integer> visited = new HashSet<>();
        ArrayList<ArrayList<Integer>> adjList = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            ArrayList<Integer> neighbours = new ArrayList<>();
            adjList.add(neighbours);
        }
        for (int i = 0; i < prerequisites.length; i++) {
            int course = prerequisites[i][0];
            adjList.get(course).add(prerequisites[i][1]);
        }
        for (int i = 0; i < numCourses; i++) {
            if (!dfs(adjList, visited, i)) {
                return false;
            }
        }
        return true;
        
    }

    public static boolean dfs(ArrayList<ArrayList<Integer>> adjList, HashSet<Integer> visited, int start) {
        
        if (visited.contains(start)) {
            return false;
        }
        if (adjList.get(start).size() == 0) {
            return true;
        }
        visited.add(start);
        ArrayList<Integer> preReqs = adjList.get(start);
        for(int pre: preReqs){
            if(!dfs(adjList, visited, pre))
                return false;
        }
        visited.remove(start);
        adjList.get(start).clear();
        return true;
    }
}

File: CourseSchedule.py
from typing import List

class CourseSchedule:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adjList = []
        for i in range(numCourses):
            adjList.append([])
        for prerequisite in prerequisites:
            adjList[prerequisite[0]].append(prerequisite[1])
        return not self.dfs(adjList, numCourses)

    def dfs(self, adjList, numCourses):
        visited = []
        recStack = []
        for i in range(numCourses):
            visited.append(False)
            recStack.append(False)
        for i in range(numCourses):
            if not visited[i]:
                if self.dfsRec(adjList, i, visited, recStack):
                    return True
        return False
    
    def dfsRec(self, adjList, curr, visited, recStack):
        visited[curr] = True
        recStack[curr] = True
        for neighbour in adjList[curr]:
            if (not visited[neighbour]) and self.dfsRec(adjList, neighbour, visited, recStack):
                return True
            elif recStack[neighbour]:
                return True
        recStack[curr] = False
        return False